### PR TITLE
crimson/os/seastore/rbm: write merge-able ool extents at once

### PR DESCRIPTION
**Pull request by myoungwon**
This commit intens to write meraged-able ool extents, each of which is located at consecutive address, to the ool at once.

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 1]
```diff
@@ -999,57 +999,122 @@ RandomBlockOolWriter::do_write(
...
+
+  return seastar::do_with(
+      std::vector<io_desc_t>(),
+      [this, &t, extents, FNAME](auto& io) {
+
+    for (auto& cur : extents) {
+      auto& stats = t.get_ool_write_stats();
+      stats.extents.num += 1;
+      stats.extents.bytes += cur->get_length();
+      stats.num_records += 1;
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1017]: Probably `num_records` can be increased once per batch, similar to the use case of the segmented backend.

Then we will be able to understand how merging is really working for RBM.

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 2]
```diff
@@ -999,57 +999,122 @@ RandomBlockOolWriter::do_write(
   assert(!extents.empty());
   DEBUGT("start with {} allocated extents",
          t, extents.size());
-  return trans_intr::do_for_each(extents,
-    [this, &t, FNAME](auto& ex) {
-    auto paddr = ex->get_paddr();
-    assert(paddr.is_absolute());
-    RandomBlockManager * rbm = rb_cleaner->get_rbm(paddr); 
-    assert(rbm);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1007]: Seems it is removed unexpectedly?

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 3]
```diff
@@ -999,57 +999,122 @@ RandomBlockOolWriter::do_write(
...
+      if (can_inplace_rewrite(t, cur)) {
+	assert(cur->is_logical());
+	auto r = cur->template cast<LogicalCachedExtent>()->get_modified_region();
+	ceph_assert(r.has_value());
+	extent_len_t offset = p2align(r->offset, rbm->get_block_size());
+	cur_len = p2roundup(r->offset + r->len, rbm->get_block_size()) - offset;
+	cur_base = cur->get_paddr() + offset;
+      } else {
+	auto& trans_stats = get_by_src(w_stats.stats_by_src, t.get_src());
+	++(trans_stats.num_records);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1033]: Same here.

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 4]
```diff
@@ -999,57 +999,122 @@ RandomBlockOolWriter::do_write(
...
+	io.push_back(desc);
       }
-      return alloc_write_iertr::now();
+      TRACE("current extent: base off {} len {}, previous extent: base off {} len {}",
+	cur_base, cur_len, io.back().base, io.back().len);
+    }
+
+    return trans_intr::do_for_each(io,
+      [this, &t, FNAME](auto& e) {
+	bufferptr bp;
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1064]: Nit, can the indention be adjusted to 2 spaces per nesting?

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 5]
```diff
@@ -999,57 +999,122 @@ RandomBlockOolWriter::do_write(
...
+	      ex->get_paddr() + ex->get_length() : e.base + e.len;
+	    extent_len_t cur_len = end.as_blk_paddr().get_device_off() -
+	      ex->get_paddr().as_blk_paddr().get_device_off();
+	    bp.copy_in(cursor, cur_len, ex->get_bptr().c_str());
+	    cursor += cur_len;
+	  }
+	}
+
+	return seastar::do_with(
+	  std::move(bp),
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1089]: Probably integrating `bp` into `io_desc_t` is a good idea to save a `do_with()`?

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 6]
```diff
@@ -999,57 +999,122 @@ RandomBlockOolWriter::do_write(
...
+	} else {
+	  // handle multiple extents at once
+	  extent_len_t cursor = 0;
+	  bp = ceph::bufferptr(ceph::buffer::create_page_aligned(e.len));
+	  for (auto& ex : e.consecutive_extents) {
+	    paddr_t end = e.base + e.len > ex->get_paddr() + ex->get_length() ?
+	      ex->get_paddr() + ex->get_length() : e.base + e.len;
+	    extent_len_t cur_len = end.as_blk_paddr().get_device_off() -
+	      ex->get_paddr().as_blk_paddr().get_device_off();
+	    bp.copy_in(cursor, cur_len, ex->get_bptr().c_str());
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1083]: Looks this is a tradeoff between memory copy and io merging.

**Comment by myoungwon** [comment]
@cyx1231st Can you take a look?

**Comment by cyx1231st** [comment]
jenkins retest this please

File: src/crimson/os/seastore/extent_placement_manager.h [hunk 7]
```diff
@@ -194,6 +194,7 @@ class RandomBlockOolWriter : public ExtentOolWriter {
     paddr_t offset;
     ceph::bufferptr bp;
     RandomBlockManager* rbm;
+    extent_len_t len;
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.h:197]: Looks to be an unnecessary field, length of bp is already available.

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 8]
```diff
@@ -1002,54 +1002,77 @@ RandomBlockOolWriter::do_write(
...
-        ex->template cast<LogicalCachedExtent>());
+	ex->template cast<LogicalCachedExtent>());
     } else {
       ceph_assert("impossible");
     }
+
+    // TODO : allocate a consecutive address based on a transaction
+    if (writes.size() != 0 &&
+       (writes.back().offset + writes.back().len == cur_base &&
+       writes.back().bp.c_str() + writes.back().len == ex->get_bptr().c_str())) {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1048]: Not following `writes.back().bp.c_str() + writes.back().len == ex->get_bptr().c_str()` ?

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 9]
```diff
@@ -1002,54 +1002,77 @@ RandomBlockOolWriter::do_write(
...
-    extent_len_t offset = 0;
+
+    RandomBlockManager * rbm = rb_cleaner->get_rbm(ex->get_paddr()); 
+    assert(rbm);
+    paddr_t cur_base = ex->get_paddr();
+    assert(ex_base.is_absolute());
+    extent_len_t cur_len = ex->get_length();
     bufferptr bp;
+    TRACE("write extent {}, paddr {} ...",
+	  fmt::ptr(ex.get()), cur_base);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1019]: Seems the logic above is unchanged.

Drop unnecessary/unrelated changes to save review efforts, unless the change makes it more readable.

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 10]
```diff
@@ -1002,54 +1002,77 @@ RandomBlockOolWriter::do_write(
...
     }
-    writes.push_back(write_info_t{paddr + offset, std::move(bp), rbm});
 
     if (ex->is_initial_pending()) {
       t.mark_allocated_extent_ool(ex);
     } else if (can_inplace_rewrite(t, ex)) {
       assert(ex->is_logical());
       t.mark_inplace_rewrite_extent_ool(
-        ex->template cast<LogicalCachedExtent>());
+	ex->template cast<LogicalCachedExtent>());
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1040]: Please drop unrelated change.

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 11]
```diff
@@ -1002,54 +1002,77 @@ RandomBlockOolWriter::do_write(
...
   return trans_intr::make_interruptible(
     seastar::do_with(std::move(writes),
-      [](auto& writes) {
+      [&t, this](auto& writes) {
       return crimson::do_for_each(writes,
-        [](auto& info) {
+        [&t, this](auto& info) {
+	auto& stats = t.get_ool_write_stats();
+	stats.num_records += 1;
+	auto& trans_stats = get_by_src(w_stats.stats_by_src, t.get_src());
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1074]: Nit, move stats references out of `do_for_each`, there is no need to find the stats in each repeat.

File: src/crimson/os/seastore/extent_placement_manager.cc (hunk 8 above)
Comment (myoungwon) [reply src/crimson/os/seastore/extent_placement_manager.cc]: This checks if the previous `bp`'s address and the current extent's `bp`'s address are aligned in a row. As a result,  when submitting ` rbm->write`, we only need a bp without performing a memory copy. If we allocate consecutive memory blocks based on a transaction, the likelihood of this merging will increase.

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 12]
```diff
@@ -1002,54 +1002,77 @@ RandomBlockOolWriter::do_write(
          t, extents.size());
   std::vector<write_info_t> writes;
   writes.reserve(extents.size());
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/extent_placement_manager.cc]: Right.

File: src/crimson/os/seastore/extent_placement_manager.h (hunk 7 above)
Comment (myoungwon) [reply src/crimson/os/seastore/extent_placement_manager.h]: Removed

File: src/crimson/os/seastore/extent_placement_manager.cc (hunk 9 above)
Comment (myoungwon) [reply src/crimson/os/seastore/extent_placement_manager.cc]: Done.

File: src/crimson/os/seastore/extent_placement_manager.cc (hunk 10 above)
Comment (myoungwon) [reply src/crimson/os/seastore/extent_placement_manager.cc]: Done.

File: src/crimson/os/seastore/extent_placement_manager.cc (hunk 11 above)
Comment (myoungwon) [reply src/crimson/os/seastore/extent_placement_manager.cc]: Done.

**Comment by myoungwon** [comment]
@cyx1231st Please take another look.

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 13]
```diff
@@ -1043,11 +1039,36 @@ RandomBlockOolWriter::do_write(
...
       ceph_assert("impossible");
     }
+
+    // TODO : allocate a consecutive address based on a transaction
+    if (writes.size() != 0 &&
+       (writes.back().offset + writes.back().bp.length() == paddr &&
+       writes.back().bp.c_str() + writes.back().bp.length() ==
+       ex->get_bptr().c_str())) {
+      // We can write both the currrent extent and the previous one at once
+      // if the extents are located in a raw
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1049]: raw -> row.

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 14]
```diff
@@ -1043,11 +1039,36 @@ RandomBlockOolWriter::do_write(
...
+      writes.back().bp = buffer::claim_char(
+	writes.back().bp.length() + ex->get_bptr().length(),
+	writes.back().bp.c_str());
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.cc:1052]: So this means the life cycle of of `writes.back().bp` is relying on that both extents must be alive during the followup writes, considering the transaction can be invalidated in the middle.

File: src/crimson/os/seastore/extent_placement_manager.cc (hunk 8 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/extent_placement_manager.cc]: The precondition of this optimization seems too strict:
1. The two extents must be consecutive in `extents`
2. Their memory addresses of the buffers should be consecutive
3. Their physical addresses should be consecutive 

If I understand the "TODO" correctly, seems to me 2 is unlikely until "TODO" is working correctly. How about using memory copy to replace 2 for now?

Comment (myoungwon) [reply src/crimson/os/seastore/extent_placement_manager.cc]: Will do.

File: src/crimson/os/seastore/extent_placement_manager.cc (hunk 14 above)
Comment (myoungwon) [reply src/crimson/os/seastore/extent_placement_manager.cc:1052]: Right.

File: src/crimson/os/seastore/extent_placement_manager.cc [hunk 15]
```diff
@@ -1043,11 +1039,36 @@ RandomBlockOolWriter::do_write(
...
+    } else {
+      // Write a single extent in the existing way
+      write_info_t w_info;
+      w_info.offset = paddr;
+      w_info.rbm = rbm;
+      w_info.bp = bp;
+      writes.push_back(w_info);
+    }
+    TRACE("current extent: base off {} len {}, previous extent: base off {} len {}",
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/extent_placement_manager.cc:1061]: Right. I'll fix this.

**Comment by myoungwon** [comment]
jenkins test make check

**Comment by myoungwon** [comment]
@cyx1231st I addressed your comments.

**Review by cyx1231st — APPROVED** [review]
Approving the optimization.

But I'm not sure how much this change really helps. It would be helpful to attach performance results before and after an optimization (even if negligible), or any other useful analysis related. They are critical to understand/evaluate whether the direction/efforts are promising/worthwhile.

**Comment by myoungwon** [comment]
@cyx1231st Latency for `do_write() ` is improved from 0.8ms to 0.6ms on average. But, there's only a slight increase in IOPS—about 1% higher than before. I think I need to add the case 2 (memory aligned) you mentioned.

**Comment by cyx1231st** [comment]
> I think I need to add the case 2 (memory aligned) you mentioned.

Might be good to see how it may help or not as a PoC.

> // TODO : allocate a consecutive address based on a transaction

For RBM, reserve both consecutive paddrs and buffer memory for a transaction -- I think they are connected and can be introduced together.

Let me know whether this PR needs further updates. I'll proceed to validate this PR

**Comment by cyx1231st** [comment]
https://pulpito.ceph.com/yingxin-2024-09-04_07:41:14-crimson-rados-wip-yingxin-testing-2024-09-04-1123-distro-default-smithi/
* 7889058/76/77: https://tracker.ceph.com/issues/63844
* 7889062/63: https://tracker.ceph.com/issues/67791
* 7889088: https://tracker.ceph.com/issues/66852

No new issue found.

**Pull request merged**